### PR TITLE
[BUG] Improve default hidden layer activation in CNN based classifier and regressor

### LINKED ISSUE #9049: [BUG] sigmoid hidden activation in CNN network, classifier & regressor
**Describe the bug**

CNN Network, classifier & regressor currently has `sigmoid` as default hidden layer activation.

---

**To Reproduce**

Check [line 53](https://github.com/sktime/sktime/blob/main/sktime/networks/cnn.py#L53) in `CNNNetwork` class.
Check [line 110](https://github.com/sktime/sktime/blob/main/sktime/regression/deep_learning/cnn.py#L110) in `CNNRegressor` class.
Check [line 103](https://github.com/sktime/sktime/blob/main/sktime/classification/deep_learning/cnn.py#L103) in `CNNClassifier` class.

---

**Expected behaviour & Additional context**

In the hidden layers of a CNN, it is non-standard to use `sigmoid` as an activation. Sigmoid outputs are in (0,1) and not zero-centred, which can slow convergence. Also, in general sigmoid is used a final layer activation in binary classification. 

`ReLU` or a variation of ReLU is a generally agreeable choice for activation in hidden layers of a CNN.

---

**Versions**
<details>

System:
    python: 3.12.10 (main, Oct 30 2025, 20:57:26) [Clang 17.0.0 (clang-1700.3.19.1)]
executable: /.pyenv/versions/sktime-dev-py-3.12.10/bin/python
   machine: macOS-15.6.1-arm64-arm-64bit

Python dependencies:
          pip: 25.3
       sktime: 0.39.0
      sklearn: 1.6.1
       skbase: 0.12.6
        numpy: 2.1.3
        scipy: 1.15.3
       pandas: 2.3.3
   matplotlib: 3.10.7
       joblib: 1.5.2
huggingface-hub: None
        numba: 0.62.1
     pmdarima: None
pytorch-forecasting: None
   skforecast: 0.18.0
        skpro: 2.9.4
statsforecast: 2.0.3
  statsmodels: 0.14.5
 transformers: None
      tsfresh: None
      tslearn: None
        torch: 2.9.0
   tensorflow: 2.19.1

</details>

### PR DESCRIPTION
**Pull request by RecreationalMath**
#### Fixes #9049  

---

#### What does this implement/fix? Explain your changes.

Changed default hidden layer activation from `sigmoid` to `relu` in `CNNNetwork`, `CNNClassifier` & `CNNRegressor` class.

---

#### Does your contribution introduce a new dependency? If yes, which one?
No.

---

#### What should a reviewer concentrate their feedback on?
If `relu` makes sense for default value of activation in hidden layers in CNN based classifier and regressor.

---

#### Did you add any tests for the change?
Not required.

---

#### Any other comments?

In the hidden layers of a CNN, it is non-standard to use sigmoid as an activation. Sigmoid outputs are in (0,1) and not zero-centred, which can slow convergence. Also, in general sigmoid is used a final layer activation in binary classification.

ReLU or a variation of ReLU is a generally agreeable choice for activation in hidden layers of a CNN.

---

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Comment by yarnabrina** [comment]
FYI @noxthot @achieveordie if you have thoughts/opinions regarding this.

(As with the previous PR, I'll still note the technically breaking and deprecation requirement points at least for consideration.)

**Comment by noxthot** [comment]
I agree that `ReLU` is usually used for hidden layers nowadays since sigmoid also suffers from vanishing gradients.
Note that this is a change in behavior though.

**Review by fkiraly — CHANGES_REQUESTED** [review]
This is a change in default behaviour, yes - this would need deprecation if we do that. Can you explain why you would recommend this?

Could you kindly follow this playbook for parameter change?
https://www.sktime.net/en/stable/developer_guide/deprecation.html#changing-the-default-value-of-a-parameter

**Comment by RecreationalMath** [comment]
> This is a change in default behaviour, yes - this would need deprecation if we do that. Can you explain why you would recommend this?

For CNNs processing time series data, ReLU is the overwhelming standard choice for hidden layers, just as it is for image CNNs.

1. `ReLU` helps gradients flow better as it is differentiable to 1 (for positive values), whereas max value of `sigmoid`'s is 0.25. This consistently less than `0.25` value of the differentiable leads to vanishing gradients.
2. There is no explicit requirement of a bounded activation in the hidden layers of a CNN, unlike the final layer of a classifier. So why compress within a small range, that too, between 0 and 1?
3. `sigmoid` is computationally heavy, `ReLU` is a simple operation. This leads to `ReLU` being computationally efficient and much faster.
4. The original paper [Zhao et al. 2017](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7870510), which this implementation refers, does not specify any special need/advantage of using `sigmoid` activation in the hidden layers, they mention it as a commonly used example in step 1 of section "2.2 Training of CNN". Since then this commonly used example has become less commonly used and been replaced with `ReLU` or ReLU-like functions (LeakyReLU, SELU, etc) in CNNs for the above mentioned reasons.
5. There are other advantages that I can think for using `ReLU` here, but I'm interested in knowing if there are any time-series specific advantages that `sigmoid` provides over `ReLU` as a default that I am not aware of?

**Comment by RecreationalMath** [comment]
> (As with the previous PR, I'll still note the technically breaking and deprecation requirement points at least for consideration.)

I have replied [here](https://github.com/sktime/sktime/pull/9048#issuecomment-3509161083) in the comment on the respective PR about this.

**Comment by RecreationalMath** [comment]
> Could you kindly follow this playbook for parameter change? https://www.sktime.net/en/stable/developer_guide/deprecation.html#changing-the-default-value-of-a-parameter

Sure, I have added the deprecation logic and pushed the change to PR. Do check if the version numbers are mentioned correctly.

I have a question though.
Below I'm reproducing the [step 1 example](https://www.sktime.net/en/stable/developer_guide/deprecation.html#step-1-during-deprecation-period) mentioned on the link you shared.

```python
from sktime.utils.warnings import warn

# TODO (release <MAJOR>.<MINOR>.0)
# change the default of 'parameter' to <new_value>
# update the docstring for parameter
class EstimatorName:
    """The old docstring with deprecation info.

    Parameters
    ----------
    parameter : str, default="old_default"
        The parameter description.
        Default value of parameter will change to <new_value>
        in version '<MAJOR>.<MINOR>.0'.
    """
    def __init__(self, parameter="changing_value"):
        self.parameter = parameter
        # TODO (release <MAJOR>.<MINOR>.0)
        # change the default of 'parameter' to <new_value>
        # remove the following 'if' check
        # de-indent the following 'else' check
        if parameter == "changing_value":
            warn(
                "in `EstimatorName`, the default value of parameter 'parameter'"
                " will change to <new_value> in version '<MAJOR>.<MINOR>.0'. "
                "To keep current behaviour and to silence this warning, "
                "set 'parameter' to 'old' explicitly.",
                category=DeprecationWarning,
                obj=self,
            )
            self._parameter = "old_default"
        else:
            self._parameter = parameter

    def fit(self, X, y):
        parameter = self._parameter
        # Fit the model using parameter
        fitting_logic(parameter)
        return self

    def predict(self, X):
        parameter = self._parameter
        # Predict using the fitted model
        y_pred = prediction_logic(parameter)
        return y_pred
```
According to this implementation, the `changing_value` can effectively never be used till the `old_default`'s deprecation is not completed. Even if the user explicitly passes it in the `EstimatorName`'s construction this implementation prevents it from actually getting used and overrides it.

Essentially this implementation translates to: the new default setting (`changing_value`) is coming in version X.Y.0 but till then we won't let you use it even if you explicitly pass it in the `EstimatorName`'s construction.

That seems rather strange, doesn't it? @fkiraly

**Comment by noxthot** [comment]
> For CNNs processing time series data, ReLU is the overwhelming standard choice for hidden layers, just as it is for image CNNs.

While in general I agree with your post and conclusion, some minor comments.
 
> 1. `ReLU` helps gradients flow better as it is differentiable to 1 (for positive values), whereas max value of `sigmoid`'s is 0.25. This consistently less than `0.25` value of the differentiable leads to vanishing gradients.
The problem is not that the values are consistently less than 0.25. The problem is that the gradients are indeed approaching zero for `sigmoid(x)` with `x` going to plus or minus infinity. There is only a small range of `x` values where the gradients are sufficiently large to allow the model to converge fast.

> 5. There are other advantages that I can think for using `ReLU` here, but I'm interested in knowing if there are any time-series specific advantages that `sigmoid` provides over `ReLU` as a default that I am not aware of?

I am not a maintainer of this package. My concern would solely be the change of behaviour for existing users who have a system running with default values for the activation function. While I agree that `ReLU` is faster and does not suffer from vanishing gradients for positive values, it actually is setting gradients to zero for negative values. Which means that users who normalized their data (center around zero and standard deviation 1) will probably get worse performance when suddenly the hidden layers are changed for `ReLU`. So while I would welcome this change, I would definitely do so following the deprecation guidelines.

Concluding I would like to state that while setting `ReLU` is common nowadays, and setting `sigmoid` for the hidden layers is rather not common anymore to the best of my knowledge, it worked just fine for many years with some problems attached. `ReLU` is not superior in every aspect and introduces new problems/things to look out for. My personal choice would be leaky ReLU since it does not kill nodes if the pre-activation is negative (which might happen for unlucky preprocessing of data and choice of weight initialization).

**Comment by RecreationalMath** [comment]
> I am not a maintainer of this package. My concern would solely be the change of behaviour for existing users who have a system running with default values for the activation function. 

I've implemented the deprecation logic, so that the change happens gradually (that is, if the change is green lighted) and with enough warnings over time that, hopefully, any user's current working setup doesn't break.

> Concluding I would like to state that while setting `ReLU` is common nowadays, and setting `sigmoid` for the hidden layers is rather not common anymore to the best of my knowledge, it worked just fine for many years with some problems attached. `ReLU` is not superior in every aspect and introduces new problems/things to look out for. My personal choice would be leaky ReLU since it does not kill nodes if the pre-activation is negative (which might happen for unlucky preprocessing of data and choice of weight initialization).

I agree that the most appropriate activation function would perhaps be task/data/setup dependent and everyone will have different variations and preferences, there really is no free lunch. It's all tradeoffs. 

I am also rather new to this package, and appreciate your thoughtful comments. While I do advocate for the most commonly used options to slowly become the defaults, because the standards which come into practice organically, over time, usually suggest that they work for most cases and serve as good starting points. And that is precisely how [Zhao et al. 2017](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7870510) also made the choice. But I am happy to go with whatever the consensus comes out to be.

**Comment by RecreationalMath** [comment]
> > Could you kindly follow this playbook for parameter change? https://www.sktime.net/en/stable/developer_guide/deprecation.html#changing-the-default-value-of-a-parameter
> 
> Sure, I have added the deprecation logic and pushed the change to PR. Do check if the version numbers are mentioned correctly.
> 
> I have a question though. Below I'm reproducing the [step 1 example](https://www.sktime.net/en/stable/developer_guide/deprecation.html#step-1-during-deprecation-period) mentioned on the link you shared.
> 
> ```python
> from sktime.utils.warnings import warn
> 
> # TODO (release <MAJOR>.<MINOR>.0)
> # change the default of 'parameter' to <new_value>
> # update the docstring for parameter
> class EstimatorName:
>     """The old docstring with deprecation info.
> 
>     Parameters
>     ----------
>     parameter : str, default="old_default"
>         The parameter description.
>         Default value of parameter will change to <new_value>
>         in version '<MAJOR>.<MINOR>.0'.
>     """
>     def __init__(self, parameter="changing_value"):
>         self.parameter = parameter
>         # TODO (release <MAJOR>.<MINOR>.0)
>         # change the default of 'parameter' to <new_value>
>         # remove the following 'if' check
>         # de-indent the following 'else' check
>         if parameter == "changing_value":
>             warn(
>                 "in `EstimatorName`, the default value of parameter 'parameter'"
>                 " will change to <new_value> in version '<MAJOR>.<MINOR>.0'. "
>                 "To keep current behaviour and to silence this warning, "
>                 "set 'parameter' to 'old' explicitly.",
>                 category=DeprecationWarning,
>                 obj=self,
>             )
>             self._parameter = "old_default"
>         else:
>             self._parameter = parameter
> 
>     def fit(self, X, y):
>         parameter = self._parameter
>         # Fit the model using parameter
>         fitting_logic(parameter)
>         return self
> 
>     def predict(self, X):
>         parameter = self._parameter
>         # Predict using the fitted model
>         y_pred = prediction_logic(parameter)
>         return y_pred
> ```
> 
> According to this implementation, the `changing_value` can effectively never be used till the `old_default`'s deprecation is not completed. Even if the user explicitly passes it in the `EstimatorName`'s construction this implementation prevents it from actually getting used and overrides it.
> 
> Essentially this implementation translates to: the new default setting (`changing_value`) is coming in version X.Y.0 but till then we won't let you use it even if you explicitly pass it in the `EstimatorName`'s construction.
> 
> That seems rather strange, doesn't it? @fkiraly

Had a discussion with Franz regarding the deprecation example and opened #9060 to improve the wording of the section.

**Comment by RecreationalMath** [comment]
> In order to clarify what algorithm is in the Zhao paper, the docstrings should be added a comment at the end that, in order to obtain the Zhao algorithm, the value `"sigmoid"` should be used.

Currently the default still is `sigmoid` and will remain so till version `0.41.0`. The need to add the requested line would arrise once the default changes from `sigmoid` to `something_else`, i.e. in version `0.41.0`. 

I suggest we mention this line's addition as a sub-point in the `TODO` comments so that it doesn't get missed in version `0.41.0`.

What do you suggest - adding it in `TODO` comments or directly adding it to the docstring?

**Comment by fkiraly** [comment]
Yes, strictly speaking this would result in a todo - but a statement along the lines of "Zhao use sigmoid for activation" is already true, although redundant. Adding it now reduces the number of todos and hence the complexity of the release actions, so that is what I would go with.

**Comment by fkiraly** [comment]
can you add the line to the docstring?

**Comment by RecreationalMath** [comment]
> Yes, strictly speaking this would result in a todo - but a statement along the lines of "Zhao use sigmoid for activation" is already true, although redundant. Adding it now reduces the number of todos and hence the complexity of the release actions, so that is what I would go with.

Ok, have modified the DOCSTRING in `CNNClassifier`, `CNNRegressor` & `CNNNetwork` to specify that Zhao et al. uses `sigmoid` as hidden activation and to obtain Zhao et al. behaviour same should be used.

**Review by fkiraly — APPROVED** [review]
Thanks!

We did not release in 0.40.0, so I am moving the target version to 0.42.0, from 0.41.0
